apps/arweave/src/ar_pool.erl

```diff
@@ -0,0 +1,764 @@
+terminate(_Reason, _State) ->
+	ok.
+
+%%%===================================================================
+%%% Private functions.
+%%%===================================================================
+
+get_jobs(PrevOutput, SessionKeys, JobCache) ->
+	case SessionKeys of
+		[] ->
+			#jobs{};
+		[{NextSeed, Interval, NextVDFDifficulty} = SessionKey | _] ->
+			Jobs = maps:get(SessionKey, JobCache),
+			{Seed, Diff, Jobs2} = collect_jobs(Jobs, PrevOutput, ?GET_JOBS_COUNT),
+			Jobs3 = [#job{ output = O, global_step_number = SN,
+					partition_upper_bound = U } || {O, SN, U} <- Jobs2],
+			#jobs{ vdf = Jobs3, seed = Seed, diff = Diff, next_seed = NextSeed,
+					interval_number = Interval, next_vdf_difficulty = NextVDFDifficulty }
+	end.
+
+collect_jobs([], _PrevO, _N) ->
+	{<<>>, 0, []};
+collect_jobs(_Jobs, _PrevO, 0) ->
+	{<<>>, 0, []};
+collect_jobs([{O, _SN, _U, _S, _Diff} | _Jobs], O, _N) ->
+	{<<>>, 0, []};
+collect_jobs([{O, SN, U, S, Diff} | Jobs], PrevO, N) ->
+	{S, Diff, [{O, SN, U} | collect_jobs(Jobs, PrevO, N - 1, Diff)]}.
+
+collect_jobs([], _PrevO, _N, _Diff) ->
+	[];
+collect_jobs(_Jobs, _PrevO, 0, _Diff) ->
+	[];
+collect_jobs([{O, _SN, _U, _S, _Diff} | _Jobs], O, _N, _Diff2) ->
+	[];
+collect_jobs([{O, SN, U, _S, Diff} | Jobs], PrevO, N, Diff) ->
+	[{O, SN, U} | collect_jobs(Jobs, PrevO, N - 1, Diff)];
+collect_jobs(_Jobs, _PrevO, _N, _Diff) ->
+	%% Diff mismatch.
+	[].
+
+process_partial_solution(Solution, Ref) ->
+	PoA1 = Solution#mining_solution.poa1,
+	PoA2 = Solution#mining_solution.poa2,
+	case ar_block:validate_proof_size(PoA1) andalso ar_block:validate_proof_size(PoA2) of
+		true ->
+			process_partial_solution_field_size(Solution, Ref);
+		false ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> }
+	end.
+
+process_partial_solution_field_size(Solution, Ref) ->
+	#mining_solution{
+		nonce_limiter_output = Output,
+		seed = Seed,
+		next_seed = NextSeed,
+		mining_address = MiningAddress,
+		preimage = Preimage,
+		solution_hash = SolutionH,
+		poa1 = #poa{ chunk = C1 },
+		poa2 = #poa{ chunk = C2 }
+	} = Solution,
+	%% We have less strict deserialization in the pool pipeline to simplify
+	%% the pool proxy implementation. Therefore, we validate the field sizes here
+	%% and return the "rejected_bad_poa" status in case of a failure.
+	case {byte_size(Output), byte_size(Seed), byte_size(NextSeed), byte_size(MiningAddress),
+			byte_size(Preimage), byte_size(SolutionH), byte_size(C1), byte_size(C2)} of
+		{32, 32, 32, 32, 32, 32, L1, L2} when L1 =< ?DATA_CHUNK_SIZE, L2 =< ?DATA_CHUNK_SIZE ->
+			%% The second chunk may be either empty or 256 KiB. ar_poa:validate/1 does
+			%% the proper verification - here we simply protect against payload size abuse.
+			%% We are not strict about the first chunk here to simplify tests.
+			process_partial_solution_poa2(Solution, Ref);
+		_ ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> }
+	end.
+
+process_partial_solution_poa2(Solution, Ref) ->
```

Comment:
This is just part of the poa2 validation, right? More happens down under `process_partial_solution_poa`?

What is this actually checking? that if recall_byte2 is not set poa2 is also not set? Can we get rid o this and do that check in`process_partial_solution_poa2`? Or maybe name it something different so it's clear that it's not actually validating poa2?

Reply:
I renamed it to process_partial_solution_poa2_size